Add Refresh Token grant support

alek-sys opened the pull request:
Just a draft for now, to run tests and discuss implementation.

alek-sys commented:
Hey @jgrandja, this one is ready for your review. A couple of things I'd like to discuss before implementing are:
- Should refresh token be re-issued when a new access token is requested? The spec says "The authorization server MAY issue a new refresh token" and I see this is quite common behaviour for other authorization servers. Should I make it default behaviour or add a configuration option for that? 
- Refresh token expiration - kind of the same question, should this be configurable? I see a comment that configuring access token expiration is a TODO item, if there'll be a separate story refresh token can also be included there.

Also there is a bit of code repetition in `*Provider` classes when issuing an access token, I'm not changing that for `OAuth2RefreshTokenAuthenticationProvider`, I believe you have refactoring in mind?

jgrandja commented:
@alek-sys 

> Should refresh token be re-issued when a new access token is requested?

Let's introduce `TokenSettings.reuseRefreshTokens(boolean)` with default `false`

> Refresh token expiration - kind of the same question, should this be configurable?

Let's introduce `TokenSettings.refreshTokenTimeToLive(Duration)` with default 60 mins. If it's set to 0 then it's disabled.

> Also there is a bit of code repetition in `*Provider` classes when issuing an access token...

I'm planning on introducing `interface OAuth2TokenIssuer`. In preparation for that, can you extract the common logic between the 2 existing providers and `OAuth2RefreshTokenAuthenticationProvider ` into a `package-private` `OAuth2TokenIssuerUtil`. This way we can flush out the interface and refactor after we merge this PR.

alek-sys commented:
Thanks @jgrandja, all done now. Feel free to review whenever you are ready.

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,266 @@
+			Instant.now().plusSeconds(10),
+			this.registeredClient.getScopes());
+
+	private final OAuth2Authorization authorization = OAuth2Authorization
```
Use `TestOAuth2Authorizations.authorization()` instead

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	@Test
+	public void supportsWhenSupportedAuthenticationThenFalse() {
```
supportsWhenSupportedAuthenticationThenFalse -> supportsWhen**Un**supportedAuthenticationThenFalse

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,266 @@
+		when(this.authorizationService.findByToken(REFRESH_TOKEN_VALUE, TokenType.REFRESH_TOKEN))
+				.thenReturn(authorization);
+
+		OAuth2AccessTokenAuthenticationToken accessTokenAuthentication =
```
Let's create `OAuth2RefreshTokenAuthenticationToken` for each test instead of using the member

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,266 @@
+	@Test
+	public void authenticateWhenRefreshTokenExistsThenReturnAuthentication() {
+		when(this.authorizationService.findByToken(REFRESH_TOKEN_VALUE, TokenType.REFRESH_TOKEN))
+				.thenReturn(authorization);
```
Prefix `this` for all members

jgrandja reviewed (changes requested):
Thanks for the PR @alek-sys ! Please see review comments.

alek-sys commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:
```diff
@@ -127,45 +117,29 @@ public Authentication authenticate(Authentication authentication) throws Authent
-		OAuth2AccessToken accessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
-				jwt.getTokenValue(), jwt.getIssuedAt(), jwt.getExpiresAt(), jwt.getClaim(OAuth2ParameterNames.SCOPE));
+		OAuth2RefreshToken refreshToken = null;
+		if (registeredClient.getTokenSettings().enableRefreshTokens()) {
```
Ouch, sorry, I read "it's disabled" as "TTL is disabled" i.e. token never expires.

alek-sys commented:
Thanks for review @jgrandja, I think I addressed all your comments now. Also to recap what we discussed privately: `enableRefreshTokens` flag stays, `refreshTokenTimeToLive` is enforced to be greater than `Duration.ZERO` so that refresh tokens always have expiry date.

jgrandja commented:
Thanks for all the updates @alek-sys ! This is now in master.
